### PR TITLE
OCPQE-17714: install OSD private cluster on GCP into shared VPC (XPN)

### PR DESCRIPTION
**Pull request by jianli-wei**
- create workflow `osd-ccs-gcp-xpn-private`, to install OSD private cluster into shared VPC on GCP
- update step `osd-ccs-cluster-provision-gcp`, to support env var `PRIVATE`
- add step `osd-ccs-cluster-private` as post-installation check of private cluster
- update steps `gcp-provision-bastionhost` & `gcp-deprovision-bastionhost`, to support using the OSD QE's project as the service project
- update steps `osd-ccs-cluster-operators-wait-ready` and `osd-ccs-conf-idp-htpasswd`, to use `${SHARED_DIR}/proxy-conf.sh` if applicable
- add periodic tests in 4.13+
```
$ grep -E "as: gcp-ocm-osd-ccs-xpn-private" ci-operator/config/ -rn -A1
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__amd64-stable.yaml:190:- as: gcp-ocm-osd-ccs-xpn-private-marketplace-f28
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__amd64-stable.yaml-191-  cron: 22 18 20 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml:945:- as: gcp-ocm-osd-ccs-xpn-private-f14
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-nightly.yaml-946-  cron: 22 3 6,22 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable.yaml:237:- as: gcp-ocm-osd-ccs-xpn-private-marketplace-f14
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable.yaml-238-  cron: 21 9 3,17 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml:905:- as: gcp-ocm-osd-ccs-xpn-private-f7
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml-906-  cron: 2 18 4,11,18,25 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-stable.yaml:237:- as: gcp-ocm-osd-ccs-xpn-private-marketplace-f14
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-stable.yaml-238-  cron: 12 1 7,23 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-ec.yaml:243:- as: gcp-ocm-osd-ccs-xpn-private-marketplace-f7
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-ec.yaml-244-  cron: 38 21 6,13,20,27 * *
--
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml:889:- as: gcp-ocm-osd-ccs-xpn-private-f7
ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.16__amd64-nightly.yaml-890-  cron: 44 0 7,14,21,30 * *
$

### CODE REVIEW

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-stable-gcp-ocm-osd-ccs-xpn-private-marketplace-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-gcp-ocm-osd-ccs-xpn-private-f14

**Comment by jianli-wei** [comment]
/assign @jianlinliu @yasun1 please help review, thanks! 

cc @lento-sun

File: ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+# Required
+GCP_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/osd-ccs-gcp.json"
+
+if [ -z "${CLUSTER_NAME}" ]; then
+    CLUSTER_NAME=$(cat "${SHARED_DIR}/cluster-name")
+fi
+CLUSTER_ID=$(cat "${SHARED_DIR}/cluster-id")
+
+set +o errexit
```
Comment (jianlinliu) [inline ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: Do we do not want the job exit as failure when `check_private` return non-zero ?

Comment (jianli-wei) [reply ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: Good question, let me confirm. Thanks!

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__amd64-stable.yaml [hunk 2]
```diff
@@ -187,6 +187,17 @@ tests:
     test:
     - chain: cucushift-installer-check-cluster-health
     workflow: osd-ccs-gcp-xpn
+- as: gcp-ocm-osd-ccs-xpn-private-marketplace-f14
```
Comment (yasun1) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__amd64-stable.yaml:190]: About the period time for the jobs, let us follow up the rules defined in https://github.com/openshift/release/pull/50376(I will record it in a document in future):
4.12. 28 28
4.13. 28 14
4.14. 14 14
4.15. 14 7
4.16. 7 7

File: ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh (hunk 1 above)
Comment (jianli-wei) [reply ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: The function `check_private` will be executed fully, and its return code would be the exit code of the step, so if the function returns non-zero, it's true the step will exit with non-zero code.

Comment (jianlinliu) [reply ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: if that, what is the purpose of this line, why not drop it?

Comment (jianli-wei) [reply ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: @jianlinliu FYI my local testing results of the logic, where if without the line the function `check_private` may not execute fully (no any log/output), it'll exit if the `grep` failed. 

**Scenario A: without `set +o errexit`**
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat osd-ccs-cluster-private-commands.sh 
#!/bin/bash
set -o nounset
set -o errexit
set -o pipefail

tmp_output1="/tmp/pub"
tmp_output2="/tmp/priv"

function logger() {
  local -r log_level=$1; shift
  local -r log_msg=$1; shift
  echo "$(date -u --rfc-3339=seconds) - ${log_level}: ${log_msg}"
}

# Post installation check for private cluster
function check_private() {
  local num_priv_ip

  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output1}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in public zone with private IP addresses."
  # Expecting both api and *.apps dns record-sets in public zone are with private IP addresses
  if [ ${num_priv_ip} -lt 2 ]; then
    logger "ERROR" "One or more api/*.apps dns record-sets in public zone are with non-private IP address."
    return 1
  fi
  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output2}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in private zone with private IP addresses."
  # Expecting all api, api-int and *.apps dns record-sets in private zone are with private IP addresses
  if [ ${num_priv_ip} -lt 3 ]; then
    logger "ERROR" "One or more api/api-int/*.apps dns record-sets in private zone are with non-private IP address."
    return 1
  fi

  logger "INFO" "check_private: Cluster DNS record-sets check passed"
  return 0
}

#set +o errexit
check_private
[jiwei@jiwei-thinkpadt14gen4 ~]$ ./osd-ccs-cluster-private-commands.sh 
[jiwei@jiwei-thinkpadt14gen4 ~]$ echo $?
1
[jiwei@jiwei-thinkpadt14gen4 ~]$ 
```
**Scenario B: with `set +o errexit`**
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat osd-ccs-cluster-private-commands.sh 
#!/bin/bash
set -o nounset
set -o errexit
set -o pipefail

tmp_output1="/tmp/pub"
tmp_output2="/tmp/priv"

function logger() {
  local -r log_level=$1; shift
  local -r log_msg=$1; shift
  echo "$(date -u --rfc-3339=seconds) - ${log_level}: ${log_msg}"
}

# Post installation check for private cluster
function check_private() {
  local num_priv_ip

  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output1}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in public zone with private IP addresses."
  # Expecting both api and *.apps dns record-sets in public zone are with private IP addresses
  if [ ${num_priv_ip} -lt 2 ]; then
    logger "ERROR" "One or more api/*.apps dns record-sets in public zone are with non-private IP address."
    return 1
  fi
  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output2}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in private zone with private IP addresses."
  # Expecting all api, api-int and *.apps dns record-sets in private zone are with private IP addresses
  if [ ${num_priv_ip} -lt 3 ]; then
    logger "ERROR" "One or more api/api-int/*.apps dns record-sets in private zone are with non-private IP address."
    return 1
  fi

  logger "INFO" "check_private: Cluster DNS record-sets check passed"
  return 0
}

set +o errexit
check_private
[jiwei@jiwei-thinkpadt14gen4 ~]$ ./osd-ccs-cluster-private-commands.sh 
2024-04-03 05:36:13+00:00 - INFO: There are 0 dns record-sets in public zone with private IP addresses.
2024-04-03 05:36:13+00:00 - ERROR: One or more api/*.apps dns record-sets in public zone are with non-private IP address.
[jiwei@jiwei-thinkpadt14gen4 ~]$ echo $?
1
[jiwei@jiwei-thinkpadt14gen4 ~]$ 
```
**The input files content:** 
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat /tmp/pub /tmp/priv
NAME                                                          TYPE  TTL  DATA
api.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.     A     60   34.67.23.216
*.apps.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.  A     30   34.171.141.22

NAME                                                           TYPE  TTL  DATA
api.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.      A     60   10.0.0.3
api-int.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.  A     60   10.0.0.3
*.apps.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.   A     30   34.171.141.22

[jiwei@jiwei-thinkpadt14gen4 ~]$ ```

Comment (jianlinliu) [reply ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh:108]: That sounds some line in the script return non-zero, which hit `errexit` when the option is enabled. 

So I think we need to locate which line cause that, and enhance the script to ensure it exit as expectation but not have to disable `errexit` option. Maybe you can `set -x` in the script to get the debug info to locate which line return non-zero value when there is no this line.

File: ci-operator/step-registry/gcp/provision/bastionhost/gcp-provision-bastionhost-commands.sh [hunk 3]
```diff
@@ -89,6 +96,9 @@ CMD="gcloud compute instances create ${bastion_name} \
 if [ -n "${ATTACH_BASTION_SA}" ]; then
   CMD="${CMD} --service-account ${ATTACH_BASTION_SA} --scopes cloud-platform"
 fi
+if [[ "${OSD_QE_PROJECT_AS_SERVICE_PROJECT}" == "yes" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/gcp/provision/bastionhost/gcp-provision-bastionhost-commands.sh:99]: It just a bastion host, I guess not have to enable `--shielded-secure-boot`

Comment (jianli-wei) [reply ci-operator/step-registry/gcp/provision/bastionhost/gcp-provision-bastionhost-commands.sh:99]: The OSD QE's project `openshift-qe-group-i-osd` has Secure Boot constraint enabled, so launch instance in that project has to enable `--shielded-secure-boot`. That's it, thanks!

Comment (jianlinliu) [reply ci-operator/step-registry/gcp/provision/bastionhost/gcp-provision-bastionhost-commands.sh:99]: Got it.

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Pull request merged**